src/tests/fuzzing/fuzzer_reader.c

```diff
@@ -77,96 +77,124 @@ void fuzz_get_chunk(sc_reader_t *reader, const uint8_t **chunk, uint16_t *chunk_
+
+	if (chunk_size >= 2) {
+		/* set the SW1 and SW2 status bytes (the last two bytes of
+		 * the response */
+		apdu->sw1 = (unsigned int)chunk[chunk_size - 2];
+		apdu->sw2 = (unsigned int)chunk[chunk_size - 1];
+		chunk_size -= 2;
+		/* set output length and copy the returned data if necessary */
+		if (chunk_size <= apdu->resplen)
+			apdu->resplen = chunk_size;
+
+		if (apdu->resplen != 0)
+			memcpy(apdu->resp, chunk, apdu->resplen);
+	} else {
+		apdu->sw1 = 0x6D;
+		apdu->sw2 = 0x00;
+		apdu->resplen = 0;
+	}
+
+	return SC_SUCCESS;
 }
 
 struct sc_reader_driver *sc_get_fuzz_driver(void)
 {
-    fuzz_ops.release = fuzz_reader_release;
-    fuzz_ops.connect = fuzz_reader_connect;
-    fuzz_ops.disconnect = fuzz_reader_disconnect;
-    fuzz_ops.transmit = fuzz_reader_transmit;
-    return &fuzz_drv;
+	fuzz_ops.release = fuzz_reader_release;
+	fuzz_ops.connect = fuzz_reader_connect;
+	fuzz_ops.disconnect = fuzz_reader_disconnect;
+	fuzz_ops.transmit = fuzz_reader_transmit;
+	return &fuzz_drv;
 }
 
 void fuzz_add_reader(struct sc_context *ctx, const uint8_t *Data, size_t Size)
 {
-    sc_reader_t	*reader;
-    struct driver_data *data;
-    char name[64] = {0};
+	sc_reader_t	*reader;
+	struct driver_data *data;
+	char name[64] = {0};
+
+	if (!(reader = calloc(1, sizeof(*reader)))
+			|| !(data = (calloc(1, sizeof(*data))))) {
+		free(reader);
+		return;
+	}
+
+	data->Data = Data;
+	data->Size = Size;
+
+	reader->driver = &fuzz_drv;
+	reader->ops = &fuzz_ops;
+	reader->drv_data = data;
+	snprintf(name, sizeof name - 1, "%zu random byte%s reader (%p)",
+			Size, Size == 1 ? "" : "s", Data);
+	reader->name = strdup(name);
+
+	reader->ctx = ctx;
+	list_append(&ctx->readers, reader);
+}
 
-    if (!(reader = calloc(1, sizeof(*reader)))
-            || !(data = (calloc(1, sizeof(*data))))) {
-        free(reader);
-        return;
-    }
+int fuzz_connect_card(sc_context_t *ctx, sc_card_t **card, sc_reader_t **reader_out,
+                      const uint8_t *data, size_t size)
+{
+	struct sc_reader *reader = NULL;
+
+	/* Erase possible readers from ctx */
+	while (list_size(&ctx->readers)) {
+		sc_reader_t *rdr = (sc_reader_t *) list_get_at(&ctx->readers, 0);
+		_sc_delete_reader(ctx, rdr);
+	}
+	if (ctx->reader_driver->ops->finish != NULL)
+		ctx->reader_driver->ops->finish(ctx);
+
+	/* Create virtual reader */
+	ctx->reader_driver = sc_get_fuzz_driver();
+	fuzz_add_reader(ctx, data, size);
+	reader = sc_ctx_get_reader(ctx, 0);
 
-    data->Data = Data;
-    data->Size = Size;
+	/* Connect card */
+	if (sc_connect_card(reader, card))
+		return SC_ERROR_INTERNAL;
 
-    reader->driver = &fuzz_drv;
-    reader->ops = &fuzz_ops;
-    reader->drv_data = data;
-    snprintf(name, sizeof name - 1, "%zu random byte%s reader (%p)",
-            Size, Size == 1 ? "" : "s", Data);
-    reader->name = strdup(name);
+	if (reader_out)
+		*reader_out = reader;
 
-    reader->ctx = ctx;
-    list_append(&ctx->readers, reader);
+	return SC_SUCCESS;
```

Comment:
I think the last time, we excluded the formatting changes until https://github.com/OpenSC/OpenSC/pull/2017 is merged, but I'm fine either way, with or without changing this here.

Reply:
I moved the formatting changes into separate commit, so can be removed.

Reply:
Indeed, sorry, removed.